clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -598,6 +618,16 @@ synchronized Integer sequenceNumber(TopicPartition topicPartition) {
         return topicPartitionBookkeeper.getPartition(topicPartition).nextSequence;
     }
 
+    /**
+     * Returns the current producer id/epoch of the given TopicPartition.
+     */
+    synchronized ProducerIdAndEpoch producerIdAndEpoch(TopicPartition topicPartition) {
+        if (!isTransactional())
+            topicPartitionBookkeeper.addPartition(topicPartition);
```

Comment:
I wonder if we could create the partition state once for all instead of having to do this here and in other places. I need to find where we could do so.

Reply:
I agree that it convolutes the code to need to check the map all over the place. I guess we could add a `getOrCreate` method or something like that.